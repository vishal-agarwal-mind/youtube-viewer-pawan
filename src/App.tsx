import { useEffect } from "react";
import _ from "lodash";
import SearchBar from "./components/search_bar";
import VideoList from "./components/video_list";
import VideoDetail from "./components/video_detail";
import Comments from "./components/comments";
import { viewVideos, onVideoSelect } from "./redux/features/videoActions";
import { connect } from "react-redux";

interface VideoStates {
  term: string;
  videos: Array<any>;
  selectedVideo: object | null
}

const API_KEY: string | undefined = process.env.REACT_APP_VIDEO_API;

/**
 * Main component of the app. It is responsible for loading the list of
 * videos and managing all other components. It user Youtube Data API v3
 * to fetch video list.
 * @component
 */
const App = (props: any) => {

  /**
   * **videoSearch** is an inner function of **App** component which is
   * responsible for fetching the list of videos.
   * @param {string} [term=liverpool] - A keyword for searching videos.
  */
  let videoSearch = (term: string) => {
    let url = new URL("https://www.googleapis.com/youtube/v3/search");
    url.searchParams.append('part', 'snippet');
    url.searchParams.append('key', (API_KEY ? API_KEY : ''));
    url.searchParams.append('q', term);
    url.searchParams.append('type', 'video');

    fetch(url.toString())
    .then(response => response.json())
    .then(data => {
      // Dispatching data to store.
      props.dispatch(viewVideos(data.items ? data.items : []));
      props.dispatch(onVideoSelect(data.items[0] ? data.items[0] : null));
    })
    .catch(err => {
      console.error(err);
    });
  }

  useEffect(() => {
    videoSearch("liverpool");
    // eslint-disable-next-line
  }, []);

  const searchVideo = _.debounce((term: string)  => {
    videoSearch(term)
  }, 300);

  return (
    <div className="flex flex-col sm:w-full lg:w-5/6 mx-auto divide-y-2">
      <div className="flex flex-wrap w-full px-4 pb-5">
        <SearchBar onSearchTermChange={searchVideo} />
      </div>
      <div className="flex sm:flex-col lg:flex-row w-full divide-x-2">
        <div className="flex flex-col w-full sm:w-full lg:w-2/3 px-4">
          <div className="w-full mt-2">
            <VideoDetail />
          </div>
          <div className="flex flex-col">
            <div className="divide-y-2">
              <div>&nbsp;</div>
              <div>
                <Comments api={API_KEY} />
              </div>
            </div>
          </div>
        </div>
        <div className="w-full sm:w-full lg:w-1/3 px-4">
          <div className="py-4 mt-2">
            <VideoList />
          </div>
        </div>
      </div>
    </div>
  );
}

const mapStateToProps = (state: VideoStates) => {
  return {
    term: state.term,
    videos: state.videos,
    selectedVideo: state.selectedVideo
  }
}

export default connect(mapStateToProps)(App);
