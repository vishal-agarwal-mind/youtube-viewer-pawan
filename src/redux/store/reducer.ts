import { VIDEOS, SELECTED_VIDEO, SEARCH_VIDEO, COMMENTS, COUNT_COMMENTS, NEXT_COMMENT_PAGE } from "../types";
import { VideoStates, CommentState } from "../interface";

const initialVideoState: VideoStates = {
    term: '',
    videos: [],
    selectedVideo: null
};

const initialCommentState: CommentState = {
    total: 0,
    comments: null,
    nextPageToken: ''
}

export const videoReducer = (state = initialVideoState, action:any) => {
    let term = state.term;
    let videos = state.videos;
    let selectedVideo = state.selectedVideo;

    switch(action.type) {
        case VIDEOS:
            return { term: term, selectedVideo: selectedVideo, videos: action.payload.videos };
        case SELECTED_VIDEO:
            return { term: term, selectedVideo: action.payload.selectedVideo, videos: videos };
        case SEARCH_VIDEO:
            return { term: action.payload.term, selectedVideo: selectedVideo, videos: videos };
        default:
            return state;
    }
};

export const commentReducer = (state = initialCommentState, action: any) => {
    let total = state.total;
    let comments = state.comments;
    let nextPageToken = state.nextPageToken;
    switch(action.type) {
        case COMMENTS:
            return { total: total, comments: action.payload.comments, nextPageToken: nextPageToken };
        case COUNT_COMMENTS:
            return { total: action.payload.total, comments: comments, nextPageToken: nextPageToken };
        case NEXT_COMMENT_PAGE:
            return { total: total, comments: comments, nextPageToken: action.payload.token };
        default:
            return state;
    }
}