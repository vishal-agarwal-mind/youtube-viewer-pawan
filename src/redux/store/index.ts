import { configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { videoReducer, commentReducer } from "./reducer";

export default configureStore({
    reducer: {
        video: videoReducer,
        comment: commentReducer
    },
    middleware: [logger]
})
