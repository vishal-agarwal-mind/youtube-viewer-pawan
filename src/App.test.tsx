import ReactDOM from "react-dom";
import { Provider } from "react-redux";
import App from "./App";
import store from "./redux/store";
import { cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";

afterEach(cleanup); // Cleanup each `Test` after rendering, to prevent `Tests` from appending after last executed `Test`.

// Created `Test` using `jest`.
it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(
    <Provider store={store}>
      <App />
    </Provider>,
    div
  );
  ReactDOM.unmountComponentAtNode(div);
});

// Snapshot test
it("matches snapshot", () => {
  const tree = renderer.create(<Provider store={store}><App /></Provider>).toJSON();
  expect(tree).toMatchSnapshot()
});
