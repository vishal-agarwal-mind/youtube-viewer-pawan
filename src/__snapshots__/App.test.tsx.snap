// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="flex flex-col sm:w-full lg:w-5/6 mx-auto divide-y-2"
>
  <div
    className="flex flex-wrap w-full px-4 pb-5"
  >
    <div
      className="w-2/3 ml-auto mr-auto h-12"
    >
      <input
        className="appearance-none block w-full bg-gray-200 text-red-700 border border-gray-200 rounded py-3 px-4 my-3 leading-tight focus:outline-none focus:bg-white focus:border-gray-500 font-mono"
        onChange={[Function]}
        placeholder="SEARCH"
        value=""
      />
    </div>
  </div>
  <div
    className="flex sm:flex-col lg:flex-row w-full divide-x-2"
  >
    <div
      className="flex flex-col w-full sm:w-full lg:w-2/3 px-4"
    >
      <div
        className="w-full mt-2"
      >
        <div
          className="font-mono p-2 w-full"
        >
          Loading Video...
        </div>
      </div>
      <div
        className="flex flex-col"
      >
        <div
          className="divide-y-2"
        >
          <div>
            Â 
          </div>
          <div>
            <div
              className="w-full p-2"
            >
              <h3
                className="font-mono mt-3"
              >
                Loading Comments...
              </h3>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="w-full sm:w-full lg:w-1/3 px-4"
    >
      <div
        className="py-4 mt-2"
      >
        <div
          className="font-mono"
        >
          Loading video list...
        </div>
      </div>
    </div>
  </div>
</div>
`;
